File: buy_sell_stock.cpp
#include "leetcode.h"

class Solution {
public:
  int maxProfit(int k, vector<int> &prices) {
    if (k == 0)
      return 0;
    if (prices.size() == 0 || prices.size() == 1)
      return 0;
    vvd(int) dp(k + 1, vector<int>(prices.size()));
    for (int i = 0; i < dp.size(); i++) {
      for (int j = 0; j < dp[i].size(); j++) {
        if (i == 0 || j == 0)
          dp[i][j] = 0;
        else {
          int x, mx = INT_MIN;
          for (int k = 0; k < j; k++) {
            x = prices[j] - prices[k] + dp[i - 1][k];
            mx = max(x, mx);
          }
          dp[i][j] = max(dp[i][j - 1], mx);
        }
      }
    }
    return dp[dp.size() - 1][prices.size() - 1];
  }
};

int main() {
  Solution obj;
  int k = 2;
  vector<int> prices = {2, 4, 1};
  cout << obj.maxProfit(k, prices);
}

File: other/buy_sell_stock.rs
use std::collections::HashMap;
struct Solution;

#[derive(PartialEq, Eq, Hash, Copy, Clone)]
enum Invested {
    Yes, No, Cooldown
}

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        Self::dp(&prices, 0, Invested::No, &mut HashMap::new())
    }
    fn dp(prices: &[i32], day: usize, invested: Invested, memo: &mut HashMap<(usize, Invested), i32>) -> i32 {
        if day == prices.len() {
            0
        } else if let Some(profit) = memo.get(&(day, invested)) {
            *profit
        } else {
            let ans = match invested {
                Invested::Yes => Self::dp(prices, day + 1, invested, memo).max(prices[day] + Self::dp(prices, day + 1, Invested::Cooldown, memo)),
                Invested::No => Self::dp(prices, day + 1, invested, memo).max(-prices[day] + Self::dp(prices, day + 1, Invested::Yes, memo)),
                Invested::Cooldown => Self::dp(prices, day + 1, Invested::No, memo)
            };
            memo.insert((day, invested), ans);
            ans
        }
    }
}

fn main() {
    let prices = vec![1,2,3,0,2];
    print!("{}", Solution::max_profit(prices));
}
